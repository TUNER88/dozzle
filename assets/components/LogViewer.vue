<template lang="html">
  <div class="is-fullheight">
    <div class="search columns is-gapless is-vcentered" v-show="showSearch">
      <div class="column">
        <p class="control has-icons-left">
          <input class="input" type="text" placeholder="Filter" ref="filter" v-model="filter" />
          <span class="icon is-small is-left"><i class="fas fa-search"></i></span>
        </p>
      </div>
      <div class="column is-1 has-text-centered">
        <button class="delete is-medium" @click="resetSearch()"></button>
      </div>
    </div>

    <ul class="events">
      <li v-for="item in filtered" class="event" :key="item.key">
        <span class="date">{{ item.date | relativeTime }}</span>
        <span class="text" v-html="colorize(item.message)"></span>
      </li>
    </ul>
    <scrollbar-notification :messages="messages"></scrollbar-notification>
  </div>
</template>

<script>
import { formatRelative } from "date-fns";
import AnsiConvertor from "ansi-to-html";
import ScrollbarNotification from "./ScrollbarNotification";

const ansiConvertor = new AnsiConvertor({ escapeXML: true });

export default {
  props: ["messages"],
  name: "LogViewer",
  components: {
    ScrollbarNotification
  },
  data() {
    return {
      showSearch: false,
      filter: ""
    };
  },
  mounted() {
    window.addEventListener("keydown", this.onKeyDown);
  },
  destroyed() {
    window.removeEventListener("keydown", this.onKeyDown);
  },
  methods: {
    onKeyDown(e) {
      if ((e.metaKey || e.ctrlKey) && e.key === "f") {
        this.showSearch = true;
        this.$nextTick(() => this.$refs.filter.focus());
        e.preventDefault();
      } else if ((e.metaKey || e.ctrlKey) && e.key === "k") {
        this.messages = [];
      } else if (e.key === "Escape") {
        this.resetSearch();
      }
    },
    resetSearch() {
      this.showSearch = false;
      this.filter = "";
    },
    colorize: function(value) {
      return ansiConvertor
        .toHtml(value)
        .replace("&lt;mark&gt;", "<mark>")
        .replace("&lt;/mark&gt;", "</mark>");
    }
  },
  computed: {
    filtered() {
      const { filter, messages } = this;

      if (filter) {
        const isSmartCase = filter === filter.toLowerCase();
        const regex = isSmartCase ? new RegExp(filter, "i") : new RegExp(filter);
        return messages
          .filter(d => d.message.match(regex))
          .map(d => ({
            ...d,
            message: d.message.replace(regex, "<mark>$&</mark>")
          }));
      }
      return messages;
    }
  },
  filters: {
    relativeTime(date) {
      return formatRelative(date, new Date());
    }
  }
};
</script>
<style scoped>
.events {
  padding: 10px;
  font-family: "Roboto Mono", monaco, monospace;
}

.event {
  font-size: 13px;
  line-height: 16px;
  word-wrap: break-word;
}

.date {
  background-color: #262626;
  color: #258ccd;
}

.is-fullheight {
  min-height: 100vh;
}

.text {
  white-space: pre-wrap;
}

.search {
  width: 350px;
  position: fixed;
  padding: 10px;
  background: rgba(50, 50, 50, 0.9);
  top: 0;
  right: 0;
  border-radius: 0 0 0 5px;
}
.delete {
  margin-left: 1em;
}

/deep/ mark {
  border-radius: 2px;
  background-color: #ffdd57;
  animation: pops 0.2s ease-out;
  display: inline-block;
}

@keyframes pops {
  0% {
    transform: scale(1.5);
  }
  100% {
    transform: scale(1.05);
  }
}
</style>
