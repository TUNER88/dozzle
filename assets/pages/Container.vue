<template lang="html">
  <log-event-source :id="id" v-slot="eventSource">
    <log-viewer :messages="eventSource.messages"></log-viewer>
  </log-event-source>
</template>

<script>
import LogEventSource from "../components/LogEventSource";
import LogViewer from "../components/LogViewer";

export default {
  props: ["id", "name"],
  name: "Container",
  components: {
    LogViewer,
    LogEventSource
  },
  metaInfo() {
    return {
      title: this.name,
      titleTemplate: "%s - Dozzle"
    };
  }
};
</script>
<style scoped></style>
