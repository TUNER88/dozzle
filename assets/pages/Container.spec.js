import EventSource from "eventsourcemock";
import { sources } from "eventsourcemock";
import { shallowMount, mount } from "@vue/test-utils";
import MockDate from "mockdate";
import Container from "./Container";
import LogViewer from "../components/LogViewer.vue";

describe("<Container />", () => {
  beforeEach(() => {
    global.BASE_PATH = "";
    global.EventSource = EventSource;
    MockDate.set("6/12/2019", 0);
    window.scrollTo = jest.fn();
  });

  afterEach(() => MockDate.reset());

  test("is a Vue instance", async () => {
    const wrapper = shallowMount(Container);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("renders correctly", async () => {
    const wrapper = mount(Container);
    expect(wrapper.element).toMatchSnapshot();
  });

  test("should connect to EventSource", async () => {
    mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    expect(sources["/api/logs/stream?id=abc"].readyState).toBe(1);
  });

  test("should close EventSource", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    wrapper.destroy();
    expect(sources["/api/logs/stream?id=abc"].readyState).toBe(2);
  });

  test("should parse messages", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    sources["/api/logs/stream?id=abc"].emitMessage({ data: `2019-06-12T10:55:42.459034602Z "This is a message."` });
    const [message, _] = wrapper.find(LogViewer).vm.messages;

    expect(message).toMatchInlineSnapshot(`
      Object {
        "date": 2019-06-12T10:55:42.459Z,
        "key": 0,
        "message": " \\"This is a message.\\"",
      }
    `);
  });

  test("should render messages", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    sources["/api/logs/stream?id=abc"].emitMessage({ data: `2019-06-12T10:55:42.459034602Z "This is a message."` });

    expect(wrapper.find("ul.events")).toMatchInlineSnapshot(`
      <ul class="events">
        <li class="event"><span class="date">today at 10:55 AM</span> <span class="text"> "This is a message."</span></li>
      </ul>
    `);
  });

  test("should render messages with color", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    sources["/api/logs/stream?id=abc"].emitMessage({
      data: `2019-06-12T10:55:42.459034602Z \x1b[30mblack\x1b[37mwhite`
    });

    expect(wrapper.find("ul.events")).toMatchInlineSnapshot(`
      <ul class="events">
        <li class="event"><span class="date">today at 10:55 AM</span> <span class="text"> <span style="color:#000">black<span style="color:#AAA">white</span></span></span></li>
      </ul>
    `);
  });

  test("should render messages with html entities", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    sources["/api/logs/stream?id=abc"].emitMessage({
      data: `2019-06-12T10:55:42.459034602Z <test>foo bar</test>`
    });

    expect(wrapper.find("ul.events")).toMatchInlineSnapshot(`
      <ul class="events">
        <li class="event"><span class="date">today at 10:55 AM</span> <span class="text"> &lt;test&gt;foo bar&lt;/test&gt;</span></li>
      </ul>
    `);
  });

  test("should render messages with filter", async () => {
    const wrapper = mount(Container, {
      propsData: { id: "abc" }
    });
    sources["/api/logs/stream?id=abc"].emitOpen();
    sources["/api/logs/stream?id=abc"].emitMessage({
      data: `2019-06-11T10:55:42.459034602Z Foo bar`
    });
    sources["/api/logs/stream?id=abc"].emitMessage({
      data: `2019-06-12T10:55:42.459034602Z This is a test <hi></hi>`
    });

    wrapper.find(LogViewer).setData({ filter: "test" });

    expect(wrapper.find("ul.events")).toMatchInlineSnapshot(`
      <ul class="events">
        <li class="event"><span class="date">today at 10:55 AM</span> <span class="text"> This is a <mark>test</mark> &lt;hi&gt;&lt;/hi&gt;</span></li>
      </ul>
    `);
  });
});
