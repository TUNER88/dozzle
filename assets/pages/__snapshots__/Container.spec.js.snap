// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> renders correctly 1`] = `
<span>
  <div
    class="is-fullheight"
  >
    <div
      class="search columns is-gapless is-vcentered"
      style="display: none;"
    >
      <div
        class="column"
      >
        <p
          class="control has-icons-left"
        >
          <input
            class="input"
            placeholder="Filter"
            type="text"
          />
           
          <span
            class="icon is-small is-left"
          >
            <i
              class="fas fa-search"
            />
          </span>
        </p>
      </div>
       
      <div
        class="column is-1 has-text-centered"
      >
        <button
          class="delete is-medium"
        />
      </div>
    </div>
     
    <ul
      class="events"
    />
     
    <button
      class="button scroll-notification is-primary"
      name="fade"
      style="display: none;"
    >
      <span
        class="icon large"
      >
        <i
          class="fas fa-chevron-down"
        />
      </span>
    </button>
  </div>
</span>
`;
